### PR TITLE
Cqc coverage add in ascwds

### PR DESCRIPTION
**Pull request by RoyPriceCS50**
# Description
Added new function to merge_coverage_data job.
Function adds 1 column called "in_ascwds" which shows 0 or 1 if row has an asc-wds establishmentid.

# How to test
Unit tests passing.
Link to successful run in Glue: https://eu-west-2.console.aws.amazon.com/gluestudio/home?region=eu-west-2#/editor/job/cqc-coverage-add-in-ascwds-merge_coverage_data_job/runs

Link to Trello: https://trello.com/c/TNTEISLq

# Developer checklist
- [x] Unit test
- [x] Linked to Trello ticket
- [ ] Documentation up to date

### CODE REVIEW

File: jobs/merge_coverage_data.py [hunk 1]
```diff
@@ -16,6 +16,9 @@
 from utils.column_names.ind_cqc_pipeline_columns import (
     PartitionKeys as Keys,
 )
+from utils.column_names.merge_coverage_data_columns import (
+    MergeCoverageDataColumns as MergeColumns,
```
Comment (MarionJHolloway) [inline jobs/merge_coverage_data.py:20]: Maybe just "Coverage Columns"? We might extend this work one day and add another job, but the column names would be heritable.

File: jobs/merge_coverage_data.py [hunk 2]
```diff
@@ -121,6 +126,17 @@ def join_ascwds_data_into_cqc_location_df(
     return merged_coverage_df
 
 
+def add_flag_for_in_ascwds(
+    merged_coverage_df: DataFrame,
+) -> DataFrame:
+    merged_coverage_df = merged_coverage_df.withColumn(
+        MergeColumns.in_ascwds,
+        F.when(F.isnull(AWPClean.establishment_id), 0).otherwise(1),
```
Comment (MarionJHolloway) [inline jobs/merge_coverage_data.py:134]: You can add these to the column values script. I can show you how that's set up and how to use it.

File: tests/unit/test_merge_coverage_data.py [hunk 3]
```diff
@@ -90,5 +93,38 @@ def test_join_ascwds_data_into_cqc_location_df(self):
...
+        )
+
+    def test_add_flag_for_in_ascwds_adds_1_column_with_given_name(self):
+        self.assertTrue(
+            MergeCoverageDataColumns.in_ascwds in self.returned_in_ascwds_df.columns
+        )
+
+        self.assertEqual(
+            len(self.returned_in_ascwds_df.columns),
+            len(self.sample_in_ascwds_df.columns) + 1,
```
Comment (MarionJHolloway) [inline tests/unit/test_merge_coverage_data.py:119]: Length of expected df? That removes the need for the magic number here

File: jobs/merge_coverage_data.py [hunk 4]
```diff
@@ -121,6 +126,17 @@ def join_ascwds_data_into_cqc_location_df(
     return merged_coverage_df
 
 
+def add_flag_for_in_ascwds(
+    merged_coverage_df: DataFrame,
+) -> DataFrame:
```
Comment (MarionJHolloway) [inline jobs/merge_coverage_data.py:131]: Our feedback from the ONS was to improve our documentation, so I'm going to start pushing for us to write doc strings for functions we create or modify. I can talk you through this if you want.

Comment (RoyPriceCS50) [reply jobs/merge_coverage_data.py:131]: Add doc strings to only my function or the whole job script?

Comment (MarionJHolloway) [reply jobs/merge_coverage_data.py:129]: Seeing as there's only 2 here, it might be worth doing both.

File: jobs/merge_coverage_data.py (hunk 2 above)
Comment (RoyPriceCS50) [reply jobs/merge_coverage_data.py]: Think i've figured this out and pushed changes in my branch. Can show it in the meeting this afternoon.

File: jobs/merge_coverage_data.py [hunk 5]
```diff
@@ -97,6 +102,20 @@ def join_ascwds_data_into_cqc_location_df(
     cqc_location_import_date_column: str,
     ascwds_workplace_import_date_column: str,
 ) -> DataFrame:
+    """
+    Joins ASC-WDS reconciliation data to CQC locations.
+
+    Takes specific columns from the cleaned CQC locations dataframe and joins specific columns from the ASC-WDS reconciliation dataframe.
```
Comment (MarionJHolloway) [inline jobs/merge_coverage_data.py:108]: All columns? Yes, the job specifies specific columns, but the function would still work the same if you changed the list of columns at the top of the script.

Also, it's worth being specific here about the fact that it aligns import dates and joins on aligned import dates and location id.

File: jobs/merge_coverage_data.py [hunk 6]
```diff
@@ -97,6 +102,20 @@ def join_ascwds_data_into_cqc_location_df(
...
+
+    Args:
+        cqc_location_df (DataFrame): A dataframe of cleaned CQC locations.
```
Comment (MarionJHolloway) [inline jobs/merge_coverage_data.py:111]: Specify any columns that are essential for the function to work (i.e. location id and cqc location import date). This is important information for anyone using the function.

File: jobs/merge_coverage_data.py [hunk 7]
```diff
@@ -97,6 +102,20 @@ def join_ascwds_data_into_cqc_location_df(
...
+        ascwds_workplace_df (DataFrame): A dataframe of ASC-WDS workplaces which includes workplaces last updated or logged into within 2 years of snapshot.
+        cqc_location_import_date_column (String): A string refering to the import date column in the clean CQC locations dataframe.
```
Comment (MarionJHolloway) [inline jobs/merge_coverage_data.py:113]: You want to be specific that this string is the **name** of the column

File: jobs/merge_coverage_data.py [hunk 8]
```diff
@@ -121,6 +140,31 @@ def join_ascwds_data_into_cqc_location_df(
...
+def add_flag_for_in_ascwds(
+    merged_coverage_df: DataFrame,
+) -> DataFrame:
+    """
+    Add a column to the merged coverage dataframe which flags if CQC location is in ASC-WDS.
+
+    When row has an ASC-WDS establishmentid then value is 1, otherwise value is 0.
+
+    Args:
+        merged_coverage_df (DataFrame): A dataframe of CQC locations with ASC-WDS columns joined via locationid.
```
Comment (MarionJHolloway) [inline jobs/merge_coverage_data.py:152]: Probably worth adding that it needs estab id

File: tests/test_file_data.py [hunk 9]
```diff
@@ -2167,6 +2167,18 @@ class MergeCoverageData:
     ]
     # fmt: on
 
+    sample_in_ascwds_rows = [
+        (None,),
+        ("1",),
+    ]
+
+    # fmt: off
```
Comment (MarionJHolloway) [inline tests/test_file_data.py:2175]: You don't need to turn the linter off for these rows, they are short anyway :)

File: utils/column_values/categorical_columns_by_dataset.py [hunk 10]
```diff
@@ -20,6 +20,7 @@
 from utils.column_names.ind_cqc_pipeline_columns import (
     IndCqcColumns as IndCQC,
 )
+from utils.column_names.coverage_columns import CoverageColumns
```
Comment (MarionJHolloway) [inline utils/column_values/categorical_columns_by_dataset.py:23]: Not needed

File: utils/column_values/categorical_columns_by_dataset.py [hunk 11]
```diff
@@ -36,6 +37,7 @@
     ContemporaryCSSR,
     ASCWDSFilledPostsSource,
     EstimateFilledPostsSource,
+    InAscwds,
```
Comment (MarionJHolloway) [inline utils/column_values/categorical_columns_by_dataset.py:40]: Not needed

**Review by MarionJHolloway — CHANGES_REQUESTED** [review]
Just a few tweaks and then it should be over the line, well done!

**Review by MarionJHolloway — APPROVED** [review]
Looks good!

**Pull request merged**